runtime/value.go

```diff
@@ -0,0 +1,210 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package runtime
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/log"
+)
+
+const (
+	// DefaultVersion is the default version of an uninitialized kv value.
+	DefaultVersion = 0
+)
+
+var (
+	// ErrCreateWatch indicates an error when creating a watch.
+	ErrCreateWatch      = errors.New("error creating watch")
+	errInitWatchTimeout = errors.New("init watch timeout")
+	errNilValue         = errors.New("nil kv value")
+)
+
+// Value is a value that can be updated during runtime.
+type Value interface {
+	// Key is the key associated with value.
+	Key() string
+
+	// Watch starts watching for value updates.
+	Watch() error
+
+	// Unwatch stops watching for value updates.
+	Unwatch()
+}
+
+// UnmarshalFn unmarshals a kv value and extracts its payload.
+type UnmarshalFn func(value kv.Value) (interface{}, error)
+
+// ProcessFn processes a value.
+type ProcessFn func(value interface{}) error
+
+// updateWithLockFn updates a value while holding a lock.
+type updateWithLockFn func(value kv.Value) error
+
+type valueStatus int
+
+const (
+	valueNotWatching valueStatus = iota
+	valueWatching
+)
+
+type value struct {
+	sync.RWMutex
+
+	key              string
+	store            kv.Store
+	opts             Options
+	log              xlog.Logger
+	unmarshalFn      UnmarshalFn
+	processFn        ProcessFn
+	updateWithLockFn updateWithLockFn
+
+	status  valueStatus
+	watch   kv.ValueWatch
+	version int
+}
+
+// NewValue creates a new value.
+func NewValue(
+	key string,
+	opts Options,
+) Value {
+	v := &value{
+		key:         key,
+		opts:        opts,
+		store:       opts.KVStore(),
+		log:         opts.InstrumentOptions().Logger(),
+		unmarshalFn: opts.UnmarshalFn(),
+		processFn:   opts.ProcessFn(),
+		version:     DefaultVersion,
+	}
+	v.updateWithLockFn = v.updateWithLock
+	return v
+}
+
+func (v *value) Key() string { return v.key }
+
+func (v *value) Watch() error {
+	v.Lock()
+	defer v.Unlock()
+
+	if v.status == valueWatching {
+		return nil
+	}
+	watch, err := v.store.Watch(v.key)
+	if err != nil {
+		return CreateWatchError{innerError: err}
+	}
+	v.status = valueWatching
+	v.watch = watch
+
+	select {
+	case <-watch.C():
+	case <-time.After(v.opts.InitWatchTimeout()):
+		err = errInitWatchTimeout
+	}
+
+	if err == nil {
+		err = v.updateWithLockFn(watch.Get())
+	}
+
+	// NB(xichen): we want to start watching updates even though
+	// we may fail to initialize the value temporarily (e.g., during
+	// a network partition) so the value will be updated when the
+	// error condition is resolved.
+	go v.watchUpdates(v.watch)
+	if err != nil {
+		return InitValueError{innerError: err}
```

Comment:
While it's not strictly necessary, I can see the users check if the type of the error is `InitValueError` and take actions accordingly, instead of checking if the type of the error is not `CreateWatchError`, which I feel is more explicit, so I'd prefer to keep it unless you feel strongly.